### PR TITLE
Request ABIs and API level from devices

### PR DESCRIPTION
**Pull request by relan**
This is a prerequisite for APK compatibility checking.

### CODE REVIEW

**Comment by mvdan** [comment]
Thanks for the patch! There are a few little things that I'll point out that will be more obvious to you as you write more Go :)

File: device.go [hunk 1]
```diff
@@ -18,6 +18,8 @@ type Device struct {
 	Product string
 	Model   string
 	Device  string
+	ABIs    []string    // Hardware architecture, e.g. [armeabi-v7a,armeabi]
+	SDK     int         // API level, e.g. 23
```
Comment (mvdan) [inline device.go]: If you run `go fmt`, these will probably not be spaced as much.

File: device.go [hunk 2]
```diff
@@ -59,6 +61,33 @@ func Devices() ([]*Device, error) {
...
+			}
+			abi2 := device.AdbGetprop("ro.product.cpu.abi2")
+			if len(abi2) == 0 {
+				device.ABIs = []string{abi}
+			} else {
+				device.ABIs = []string{abi, abi2}
+			}
+		} else {
+			device.ABIs = strings.Split(abilist, ",")
+		}
```
Comment (mvdan) [inline device.go]: I would move all of this into a separate func, like `device.ABIs = getAbis(device)`

File: device.go [hunk 3]
```diff
@@ -59,6 +61,33 @@ func Devices() ([]*Device, error) {
...
+
+		// API level
+		device.SDK, err =
+				strconv.Atoi(device.AdbGetprop("ro.build.version.sdk"))
+		if err != nil {
+			continue
```
Comment (mvdan) [inline device.go]: `return nil, err`

File: device.go [hunk 4]
```diff
@@ -59,6 +61,33 @@ func Devices() ([]*Device, error) {
...
 		}
+
+		// Android 5.0 and later specify a list of ABIs
+		abilist := device.AdbGetprop("ro.product.cpu.abilist")
+		if len(abilist) == 0 {
+			// Older Android versions specify one primary ABI and
+			// optionally one secondary ABI
+			abi := device.AdbGetprop("ro.product.cpu.abi")
+			if len(abi) == 0 {
+				continue
```
Comment (mvdan) [inline device.go]: This skips the entire device. You probably want an `else` or something else.

**Comment by mvdan** [comment]
Also, I would be much happier if this ran `adb shell getprop` once and then kept a dictionary in memory briefly, which we then used to query all the values we need. This isn't too important now that we only call `adb shell getprop foo.bar` a few times, but we'll be adding more and more calls later on.

I'm thus fine with merging this, and I'll do the rewrite myself later - or you can do it yourself.

File: device.go (hunk 1 above)
Comment (mvdan) [inline device.go]: I would probably rename `SDK` to `APILevel` and remove the two comments

**Comment by relan** [comment]
> I'm thus fine with merging this, and I'll do the rewrite myself later - or you can do it yourself.

I'd prefer to refine the code until it looks really good. So please don't merge it now, we aren't in a hurry.

What would be a better behavior when `adb shell getprop` fails? I can't think of anything better than failing the whole device initialization.

**Comment by mvdan** [comment]
> I'd prefer to refine the code until it looks really good. So please don't merge it now, we aren't in a hurry.

Sure. Ping me when you have an updated version.

> What would be a better behavior when adb shell getprop fails? I can't think of anything better than failing the whole device initialization.

The device is connected, so it shouldn't fail. If it does, erroring and stopping the entire application seems like an OK solution to me.

Alternatives that pop to mind:

* Retry, e.g. up to 3 times
* Have a catch-all fallback with what's most common, e.g. api level 23 and armv7/arm abis.

File: device.go [hunk 5]
```diff
@@ -74,10 +91,52 @@ func (d *Device) AdbShell(args ...string) *exec.Cmd {
...
+		return nil, err
+	}
+	props := make(map[string]string)
+	scanner := bufio.NewScanner(stdout)
+	for scanner.Scan() {
+		// [property]: [value]
+		entry := strings.Split(scanner.Text(), ": ")
+		if len(entry) == 2 {
+			props[strings.Trim(entry[0], "[]")] = strings.Trim(entry[1], "[]")
+		}
```
Comment (mvdan) [inline device.go]: I would do this with a regex - something like `^\[\(.*\)\]: \[\(.*\)\]$'

File: device.go [hunk 6]
```diff
@@ -74,10 +91,52 @@ func (d *Device) AdbShell(args ...string) *exec.Cmd {
...
+	return props, nil
+}
+
 func getFailureCode(r *regexp.Regexp, line string) string {
 	return r.FindStringSubmatch(line)[1]
 }
 
+func getAbis(props map[string]string) []string {
+	// Android 5.0 and later specify a list of ABIs
+	abilist, exists := props["ro.product.cpu.abilist"]
```
Comment (mvdan) [inline device.go]: the common thing is to use `foo, e := ...`, so `e` instead of `exists`

File: device.go [hunk 7]
```diff
@@ -74,10 +91,52 @@ func (d *Device) AdbShell(args ...string) *exec.Cmd {
...
+	if exists {
+		return strings.Split(abilist, ",")
+	}
+
+	// Older Android versions specify one primary ABI and optionally
+	// one secondary ABI
+	abi, exists := props["ro.product.cpu.abi"]
+	if !exists {
+		return []string{}
```
Comment (mvdan) [inline device.go]: `return nil` to avoid an allocation

File: device.go [hunk 8]
```diff
@@ -74,10 +91,52 @@ func (d *Device) AdbShell(args ...string) *exec.Cmd {
...
+	}
+	abi2, exists := props["ro.product.cpu.abi2"]
+	if !exists {
+		return []string{abi}
+	} else {
```
Comment (mvdan) [inline device.go]: You can drop this else.

**Comment by mvdan** [comment]
I noticed you updated the PR, so I left some more comments. It looks much better now!

**Comment by relan** [comment]
Updated.

I'm not a big fan of regular expressions. I've rewritten that square brackets trimming line in a more readable way (I think).

**Pull request merged**